usl_pipeline/cloud_functions/main.py

```diff
@@ -336,7 +336,7 @@ def _build_feature_matrix(
             chunk_metadata = metastore.StudyAreaSpatialChunk.get_if_exists(
                 firestore.Client(), study_area_name, chunk_name
             )
-            if chunk_metadata is not None:
+            if chunk_metadata is not None and chunk_metadata.state is not None:
```

Comment:
I think it'd be good to check the value of the state rather than just that it isn't None, e.g. if the state was error, we may want to try again.

Why is this check necessary? 
If the functions are idempotent, it should be safe to re-build the chunk a second time.
If they aren't, I think we'd still have race conditions on checking the state and could still have multiple runners. It would need to be a transactional check & set to ensure one runner at a time.

How does this interact with retries? If one runner sets the state as StudyAreaChunkState.FEATURE_MATRIX_PROCESSING, then dies, would we never be able to re-run?

Reply:
With my change we have 3 possible values for state:
- None: appears when empty chunk meta object is created few lines below this comment or if the error happened during the unscaled feature matrix building logic in this same CF (but before it finishes with setting the FEATURE_MATRIX_PROCESSING state)
- FEATURE_MATRIX_PROCESSING: CF building unscaled feature matrix ends with this state. If this state appears it means that we no longer need to rerun this CF, it succeeded (no matter if retry logic thinks that there was any deadline exceeded for CF execution time or any other reasons).
- FEATURE_MATRIX_READY: It means that rescaling CF succeeded and we're good to go with this chunk being done completely.

I would love to have INIT state instead of None. But the problem with INIT state is that the logic creating initial empty object shouldn't override existing state that might be set in parallel by CF trying to set FEATURE_MATRIX_PROCESSING for instance. So it means we might need to set INIT state in transaction where we make sure the state is not set yet which is just more complicated logic than keeping the state being None initially.